{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "introductory-witness",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess, tempfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "polyphonic-outline",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "submitted\n",
      "submitted\n",
      "submitted\n",
      "submitted\n",
      "submitted\n",
      "submitted\n",
      "submitted\n",
      "submitted\n"
     ]
    }
   ],
   "source": [
    "# Submit all processes\n",
    "processes = []\n",
    "for _ in range(8):\n",
    "    # Open temp file handle to stash output\n",
    "    f = tempfile.TemporaryFile()\n",
    "    # Run executable -- will spawn child process on new thread\n",
    "    p = subprocess.Popen(['/home/dmf32881/code/vaccines/stress.sh'],stdout=f)\n",
    "    processes.append((p, f))\n",
    "    # Sanity check\n",
    "    print('submitted')\n",
    "\n",
    "# Handle output (synchronously) when processes complete\n",
    "for p, f in processes:\n",
    "    # Wait for process to complete before handling output\n",
    "    p.wait()\n",
    "    # Rewind to the beginning of the file so it can be dumped\n",
    "    f.seek(0)\n",
    "    # Close file handle\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "herbal-server",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'codeRNA'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-1d2826cd15b8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mtensorflow\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mtf\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 19\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mcodeRNA\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcodon_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhive_counts\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mget_codon_frequencies\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     20\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mcodeRNA\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcodon_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcodon_tables\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mget_codon_table\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mcodeRNA\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrna\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrna\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mRNA\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'codeRNA'"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Evolutionary algorithm for optimizing RNA sequences.\n",
    "The algorithm iteratively improves the initial population of candidate sequences by applying genetic operators of\n",
    "mutation and recombination.\n",
    "\n",
    "See: https://www.tensorflow.org/probability/api_docs/python/tfp/optimizer/differential_evolution_minimize\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "import random\n",
    "import timeit\n",
    "from datetime import timedelta\n",
    "import numpy as np\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\", category=DeprecationWarning)\n",
    "import tensorflow_probability as tfp\n",
    "import tensorflow as tf\n",
    "\n",
    "from codeRNA.codon_data.hive_counts import get_codon_frequencies\n",
    "from codeRNA.codon_data.codon_tables import get_codon_table\n",
    "from codeRNA.rna.rna import RNA\n",
    "\n",
    "\n",
    "class EvolutionaryRNAOptimizer(object):\n",
    "    def __init__(self, protein_sequence, scoring_function, population_size=50,\n",
    "                 organism=\"human\", codon_table_name=\"standard\"):\n",
    "        \"\"\"Create an EvolutionaryRNAOptimizer object.\n",
    "\n",
    "        Example scoring function (minimizing uridine content):\n",
    "        lambda seq: RNA(seq).get_uridine_content()\n",
    "\n",
    "        :param protein_sequence: str, protein sequence\n",
    "        :param scoring_function: function used to calculate scores for individual sequences (to be minimized by the\n",
    "            evolutionary algorithm)\n",
    "        :param population_size: int, the size of the population to evolve, must be larger than 4\n",
    "        :param organism: str, target organism, defaults to \"human\"\n",
    "        :param codon_table_name: str, name of the codon table, defaults to \"standard\"\n",
    "        \"\"\"\n",
    "        self.protein_sequence = protein_sequence\n",
    "        self.scoring_function = scoring_function\n",
    "        self.organism = organism\n",
    "        self.codon_table_name = codon_table_name\n",
    "        self.population_size = population_size\n",
    "        self.codon_map = self._get_codon_map()\n",
    "        self.initial_population_tensor = self._generate_initial_population()\n",
    "        self.optimization_results = None\n",
    "\n",
    "    def optimize(self, max_iterations=500, differential_weight=0.01, crossover_prob=0.025, seed=None):\n",
    "        \"\"\"Run evolutionary algorithm.\n",
    "        The results can be checked by running get_ranked_optimized_sequences().\n",
    "\n",
    "        :param max_iterations: int, maximum  number of iterations to evolve the population for\n",
    "        :param differential_weight: float, parameter controlling strength of mutation. Must be positive and less than 2\n",
    "        :param crossover_prob: float, the probability of recombination per site; must be between 0 and 1\n",
    "        :param seed: int, random seed. If None, no seed is applied\n",
    "        \"\"\"\n",
    "        start = timeit.default_timer()\n",
    "        optimization_results = tfp.optimizer.differential_evolution_minimize(\n",
    "            self._objective,\n",
    "            initial_population=self.initial_population_tensor,\n",
    "            max_iterations=max_iterations,\n",
    "            differential_weight=differential_weight,\n",
    "            crossover_prob=crossover_prob,\n",
    "            seed=seed\n",
    "        )\n",
    "        elapsed_time = timeit.default_timer() - start\n",
    "        print(f\"Elapsed time: {timedelta(seconds=elapsed_time)}\")\n",
    "        self.optimization_results = optimization_results\n",
    "\n",
    "        top_seq, top_score = self.get_optimized_sequences()[0]\n",
    "        print(f\"Best score after {max_iterations}: {top_score:.2f}\")\n",
    "\n",
    "    def get_optimized_sequences(self):\n",
    "        \"\"\" Return a ranked list of optimized sequences and their associated scores (ascending order).\n",
    "        Raises ValueError if called before optimize()\n",
    "\n",
    "        :return: list of tuples (sequence, score), sorted by score (ascending)\n",
    "        \"\"\"\n",
    "        if self.optimization_results is None:\n",
    "            raise ValueError(\"No optimized sequences available. Run optimize() first\")\n",
    "        nseqs = self._convert_to_nseqs(self.optimization_results.final_population)\n",
    "        scores = list(self.optimization_results.final_objective_values.numpy())\n",
    "        sorted_nseqs = sorted(zip(nseqs, scores), key=lambda x: x[1])\n",
    "        return sorted_nseqs\n",
    "\n",
    "    def get_initial_sequences(self):\n",
    "        \"\"\" Return a ranked list of initial population sequences (not the optimized ones!) and their associated scores\n",
    "        (ascending order).\n",
    "\n",
    "        :return: list of tuples (sequence, score), sorted by score (ascending)\n",
    "        \"\"\"\n",
    "        nseqs = self._convert_to_nseqs(self.initial_population_tensor)\n",
    "        scores = list(self._objective(self.initial_population_tensor).numpy())\n",
    "        sorted_nseqs = sorted(zip(nseqs, scores), key=lambda x: x[1])\n",
    "        return sorted_nseqs\n",
    "\n",
    "    def _generate_initial_population(self):\n",
    "        \"\"\"\n",
    "        Assembles initial population of random RNA sequences coding for given protein sequence.\n",
    "        Codon usage reflects the fequencies from the chosen species.\n",
    "\n",
    "        :return: initial population (Tensor objects to be used by the alg, not actual sequences)\n",
    "        \"\"\"\n",
    "        initial_population = []\n",
    "        for i in range(self.population_size):\n",
    "            member = []\n",
    "            for aa in self.protein_sequence:\n",
    "                #  random.choices function: https://pynative.com/python-weighted-random-choices-with-probability/\n",
    "                index_options = range(len(self.codon_map[aa][\"codons\"]))\n",
    "                index = random.choices(index_options, weights=self.codon_map[aa][\"weights\"], k=1)\n",
    "                member.extend(index)\n",
    "            initial_population.append(member)\n",
    "\n",
    "        # convert to TF object\n",
    "        initial_population = tf.convert_to_tensor((initial_population), float)\n",
    "        return initial_population\n",
    "\n",
    "    def _convert_to_nseqs(self, population):\n",
    "        \"\"\" Helper function to convert Tensor objects returned by the optimizer to RNA sequences.\n",
    "        Aka Dillion's magic hack.\n",
    "\n",
    "        :param population: population of sequences (tensor objects)\n",
    "        :return: list of str, RNA sequences\n",
    "        \"\"\"\n",
    "        # This is a hack. TF deals with continuous valued functions. We need discrete and finite.\n",
    "        # Whatever values are assigned, make them ints and take the absolute value.\n",
    "        population = np.absolute(np.array(population).astype(int))\n",
    "\n",
    "        # Now we want to do something with the values. It's possible that some values exceed the\n",
    "        # number of codons for the given position, so take the modulus. This is effectively a hashing\n",
    "        # function. It's not mathematically rigorous, but it's good enough.\n",
    "        # Finally, convert list of indices to the RNA sequence.\n",
    "        get_seq = lambda se: ''.join([self.codon_map[res]['codons'][se[i] % self._get_nc(res)]\n",
    "                                      for i, res in enumerate(self.protein_sequence)])\n",
    "        n_seqs = [get_seq(se) for se in population]\n",
    "        return n_seqs\n",
    "\n",
    "    def _get_nc(self, aa):\n",
    "        \"\"\" Helper function to get the number of codons for the given aminoacid\n",
    "\n",
    "        :param aa: str, single-letter amino acid symbol\n",
    "        :return: number of possible codons\n",
    "        \"\"\"\n",
    "        return len(self.codon_map[aa]['codons'])\n",
    "\n",
    "    def _objective(self, population):\n",
    "        \"\"\"Objective function for TF to minimize.\n",
    "\n",
    "        NOTE: TF uses gradient descent to minimize continuous valued functions.\n",
    "        The approach used here is not mathematically sound. It's a hack. But it gets the job done.\n",
    "\n",
    "        :param population: population of sequences (tensor objects)\n",
    "        :return: tf.float32, list of scores calculated for the population\n",
    "        \"\"\"\n",
    "        # Map continuous valued tensor to RNA sequence\n",
    "        n_seqs = self._convert_to_nseqs(population)\n",
    "        # Use the imported scoring function to score all sequences.\n",
    "        scores = [self.scoring_function(s) for s in n_seqs]\n",
    "        # Return TF object\n",
    "        return tf.cast(scores, float)\n",
    "\n",
    "    def _get_codon_map(self):\n",
    "        \"\"\" Helper function the get conveniently formatted dict with codon and codon frequency information per amino\n",
    "        acid.\n",
    "\n",
    "        :return: dict of dicts, maps amino acids to codons and codon frequencies in the given organism\n",
    "        \"\"\"\n",
    "        freqs = get_codon_frequencies(self.organism)\n",
    "        codon_table = get_codon_table(self.codon_table_name)\n",
    "        codon_map = dict()\n",
    "        for aa, codons in codon_table.items():\n",
    "            codon_map[aa] = {\"codons\": codons,\n",
    "                             \"weights\": [freqs[codon] for codon in codons]}\n",
    "        return codon_map\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    aa_seq = \"MEDAKNIKKGPAPFYPLEDGTAGEQLHKAMKRYALVPGTIAFTDAHIEVNITYAEYFEMSVRLAEAMKRYGLNTNHRIVVCSENSLQFFMPVLGALFIGVAVAPA\"\n",
    "    get_u_content = lambda seq: RNA(seq).get_uridine_content()\n",
    "    revo = EvolutionaryRNAOptimizer(protein_sequence=aa_seq, organism=\"human\", scoring_function=get_u_content)\n",
    "    revo.optimize(max_iterations=100)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "pressing-vacuum",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Environment (conda_tf)",
   "language": "python",
   "name": "conda_tf"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
